oeps/oep-0013.rst

```diff
@@ -0,0 +1,672 @@
+    - ``user_message`` (optional, but if provided, i18n as the expectation is
+      that this may surface in a UI)
+    - ``field_errors`` (if applicable)
+    - ``error_code`` (for example: see LMS ``has_access`` interface design)
+        *Include description in your REST API Docs.*
+
+    ::
+
+        {
+            "error_code": "course_not_started"  # a short string that the client can rely upon for handling different errors
+            "developer_message": "Verbose, plain language description of the problem for the app developer with hints about how to fix it.",
+            "user_message": "Pass this message on to the app user if needed.",
+            "field_errors": {
+                "foo": {
+                    "developer_message": "",
+                    "user_message": ""
+                }
+            }
+        }
+
+
+Versioning
+----------
+
+**Location**
+
+- Put the version in the URL so the client can see it easily when handling the
+  response logic.
+
+- Exception: Put it in the header only if it doesn't change the response
+  handling logic, like the OAuth endpoint.
+
+**Numbering**
+
+- *Must*: Versions are major only ex. ``/v0/`` ``/v1/`` ``/v2/``
+
+- *Must*: Major version should only be updated when changing the existing
+  contract in a way that breaks backwards compatibility.  The previous major
+  version should continue to be supported for old clients according to the
+  service's deprecation policy.
+
+- *Must*: Additive changes to the contract of the API, or the results, should
+  not bump the major version, unless those changes are non-optional, and break
+  existing contracts with the API.
+
+Pagination
+----------
+
+We use the conventions established by `DRF
+<http://www.django-rest-framework.org/api-guide/pagination/#pagenumberpagination>`_,
+with some extra fields:
+
+- ``next``: A URL to the next page, or null.
+- ``previous``: A URL to the previous page, or null.
+- ``count``: The total number of items.
+- ``num_pages``: The total number of pages.
+- ``results``: List of results.
+
+The query parameters are as follows:
+
+- ``page`` - the page of results (zero indexed)
+- ``page_size`` - the number of results to return per page
+- ``sort_order`` - a string indicating the desired order (no direction
+  modifiers)
+
+Documentation
+-------------
+
+Work with the Docs team to update the `edX platform API documentation
+<http://edx-platform-api.readthedocs.org/en/latest/>`_.
+
+**Format**
+
+Use `Swagger <http://swagger.io/specification/>`_ to generate API documentation
+where possible.
+
+**Docstrings**
+
+Following are our guidelines for Django Rest Framework API implementations.
+
+edX API docstrings serve two purposes:
+
+- They are interpreted by DRF to provide dynamic API documentation
+
+    e.g. https://courses.edx.org/api/user/v1/accounts/AndyA
+
+    See the `DRF documentation
+    <http://tomchristie.github.io/rest-framework-2-docs/topics/documenting-your-api>`_
```

Comment:
We use DRF 3 everywhere. This is a more relevant URL: http://www.django-rest-framework.org/topics/documenting-your-api/